### PR TITLE
database/raft: use new raft package to store core config

### PR DESCRIPTION
**Pull request by tessr**

### CODE REVIEW

**Comment by tessr** [comment]
I believe this is ready for review 😁

File: database/raft/Makefile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+
```
Comment (bobg) [inline database/raft/Makefile]: This file looks like it's for testing and debugging (and isn't required for normal building)? Please add a comment at the top that makes that clear.

File: database/raft/file.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	err = fsync(f)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	err = f.Close()
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	return errors.Wrap(os.Rename(temp, name))
```
Comment (bobg) [inline database/raft/file.go]: Nit: the important thing here is the `os.Rename` call; don't bury it inside a call to `errors.Wrap`.

File: database/raft/fsync_darwin.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+package raft
+
+import (
+	"os"
+	"syscall"
+)
+
+func fsync(f *os.File) error {
+	_, _, errno := syscall.Syscall(syscall.SYS_FCNTL, f.Fd(), syscall.F_FULLFSYNC, 0)
```
Comment (bobg) [inline database/raft/fsync_darwin.go]: Huh, can you really not rely on `f.Sync()` on all platforms?

File: database/raft/state/peer_test.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	expected_s := State{peers: map[uint64]string{}}
+
+	s.RemovePeerAddr(1)
+	if !reflect.DeepEqual(s, expected_s) {
+		t.Errorf("RemovePeerAddr(%d) => %v want %v", 1, s, expected_s)
+	}
+
+}
+
+func TestSetPeerAddr(*testing.T) {
```
Comment (bobg) [inline database/raft/state/peer_test.go]: This needs filling in or removal.

File: database/raft/state/peer_test.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+func TestGetPeerAddr(*testing.T) {
```
Comment (bobg) [inline database/raft/state/peer_test.go]: This needs filling in or removal.

File: database/raft/state/state.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	"chain/log"
+)
+
+const nextNodeID = "raft/nextNodeID"
+
+var ErrAlreadyApplied = errors.New("entry already applied")
+
+// State is a general-purpose data store designed to accumulate
+// and apply replicated updates from a raft log.
+// The zero value is an empty State ready to use.
```
Comment (bobg) [inline database/raft/state/state.go]: I don't think this is true, because the `map`s will be uninitialized (by either `make` or with a literal).

File: database/raft/state/state.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+// Provisional read operation.
+func (s *State) Get(key string) (value []byte) {
+	return s.state[key]
+}
+
+// Set encodes a set operation setting key to value.
+// The encoded op should be committed to the raft log,
+// then it can be applied with Apply.
+func Set(key string, value []byte) (instruction []byte) {
+	// TODO(kr): make a way to delete things
```
Comment (bobg) [inline database/raft/state/state.go]: Is that not `Delete`, below?

File: cmd/cored/main.go [hunk 8]
```diff
@@ -143,6 +147,55 @@ func runServer() {
 	ctx := context.Background()
 	env.Parse()
 
+	raftDir := filepath.Join(*dir, "raft") // TODO(kr): better name for this
```
Comment (bobg) [inline cmd/cored/main.go]: Yes, this does need a better name than `DIR`.

File: database/raft/fsync_darwin.go (hunk 3 above)
Comment (tessr) [inline database/raft/fsync_darwin.go]: I didn't write this line, so I'll defer to @ameets and @kr :)

File: cmd/cored/main.go (hunk 8 above)
Comment (tessr) [inline cmd/cored/main.go]: Isn't this `DIR/raft` now? But maybe we should add the raft id or something. (Not sure what makes a good directory name...)

File: database/raft/Makefile (hunk 1 above)
Comment (tessr) [inline database/raft/Makefile]: (done)

File: database/raft/fsync_darwin.go (hunk 3 above)
Comment (kr) [inline database/raft/fsync_darwin.go]: I'll defer to the explanation in [Apple's fsync man page](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man2/fsync.2.html):

     Note that while fsync() will flush all data from the host to the drive (i.e. the "permanent storage
     device"), the drive itself may not physically write the data to the platters for quite some time and it
     may be written in an out-of-order sequence.

     Specifically, if the drive loses power or the OS crashes, the application may find that only some or
     none of their data was written.  The disk drive may also re-order the data so that later writes may be
     present, while earlier writes are not.

     This is not a theoretical edge case.  This scenario is easily reproduced with real world workloads and
     drive power failures.

     For applications that require tighter guarantees about the integrity of their data, Mac OS X provides
     the F_FULLFSYNC fcntl.  The F_FULLFSYNC fcntl asks the drive to flush all buffered data to permanent
     storage.  Applications, such as databases, that require a strict ordering of writes should use F_FULLF-
     SYNC to ensure that their data is written in the order they expect.  Please see fcntl(2) for more
     detail.

(Personally I think it would be better if the default were the safe way perhaps with a flag to turn it off, and even then the flag should arguably be a system-level option rather than an API call parameter. But it is what it is.)

Linux [until about 2008 it seems](https://www.mail-archive.com/linux-fsdevel@vger.kernel.org/msg12025.html) didn't provide the safer option at all, and now I think it only provides the safer option? It's hard to get a straight answer out of the Internet.

Basically I tried to use the safest available option that I'm aware of on each platform. Incidentally, etcd itself does the same thing https://github.com/coreos/etcd/blob/2724c3946eb2f3def5ed38a127be982b62c81779/pkg/fileutil/sync_darwin.go.

Comment (kr) [inline database/raft/fsync_darwin.go]: This might be worth adding a note in the comments to explain the situation.

Comment (tessr) [inline database/raft/fsync_darwin.go]: Will do.

Comment (tessr) [inline database/raft/fsync_darwin.go]: (done)

**Comment by tessr** [comment]
This is ready for another round.

**Comment by bobg** [comment]
:eyes:

Note, the PR title should change to say `database/raft` instead of `core/raft`.

**Comment by tessr** [comment]
Good catch, thanks!

File: core/api.go [hunk 9]
```diff
@@ -97,6 +98,7 @@ func (a *API) needConfig() func(f interface{}) http.Handler {
 	return jsonHandler
 }
 
+// Adds the Core API routes to a preexisting http handler.
```
Comment (jbowens) [inline core/api.go]: Should be 
```
// Handler adds the ...
```
https://golang.org/doc/effective_go.html#commentary

File: core/config/config.proto [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+	bc.ProtoHash blockchain_id = 4;
+	string generator_url = 5;
+	string generator_access_token = 6;
+	string block_hsm_url = 7;
+	string block_hsm_access_token = 8;
+	uint64 configured_at = 9;
+	string block_pub = 10;
+	repeated BlockSigner signers = 11;
+	uint32 quorum = 12;
+	uint64 max_issuance_window = 13	;
```
Comment (jbowens) [inline core/config/config.proto]: nit: extra space before the semicolon

File: database/raft/state/state.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+	for _, op := range instr.Operations {
+		switch op.Type {
+		case statepb.Op_SET:
+			s.state[op.Key] = op.Value
+			s.version[op.Key] = index
+		case statepb.Op_DELETE:
+			delete(s.state, op.Key)
+			delete(s.version, op.Key)
+			//TODO (ameets):increment version or delete entry?
+			//s.version[op.Key] = index
```
Comment (jbowens) [inline database/raft/state/state.go]: Does this TODO need addressing?

Comment (ameets) [inline database/raft/state/state.go]: It should be deleted

File: database/raft/raft.go [hunk 12]
```diff
@@ -0,0 +1,920 @@
...
+	}
+	if crc32.Checksum(snapPB.Data, crcTable) != snapPB.Crc {
+		return errors.Wrap(errors.New("bad snapshot crc"))
+	}
+	err = snapshot.Unmarshal(snapPB.Data)
+	return errors.Wrap(err)
+}
+
+// TODO(kr): under what circumstances should we do this?
+func (sv *Service) evict(nodeID uint64) {
```
Comment (bobg) [inline database/raft/raft.go]: This stub should be removed until/unless it's needed.

File: database/raft/raft.go [hunk 13]
```diff
@@ -0,0 +1,920 @@
...
+		return nil, errors.Wrap(err)
+	}
+
+	_, st, ents, err := wal.ReadAll()
+	if err != nil {
+		return nil, errors.Wrap(err)
+	}
+
+	sv.raftStorage.ApplySnapshot(raftSnap)
+	if raft.IsEmptySnap(raftSnap) {
```
Comment (bobg) [inline database/raft/raft.go]: `if !raft.IsEmptySnap(...)` (and remove empty if branch).

File: database/raft/rand.go [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package raft
+
+import (
+	"crypto/rand"
+)
+
+func randID() []byte {
+	c := 16 // prevent collisions
```
Comment (bobg) [inline database/raft/rand.go]: Why not `var b [16]byte`?

File: cmd/corectl/main.go [hunk 15]
```diff
@@ -130,36 +142,34 @@ func configGenerator(db *sql.DB, args []string) {
 				fatalln("error:", "bad ed25519 public key length")
 			}
 			url := args[i+1]
-			signers = append(signers, config.BlockSigner{
+			signers = append(signers, &config.BlockSigner{
 				Pubkey: pubkey,
-				URL:    url,
+				Url:    url,
```
Comment (tessr) [inline cmd/corectl/main.go]: I 100% agree, but not much we can do about it :(

File: cmd/corectl/main.go [hunk 16]
```diff
@@ -8,33 +8,38 @@ import (
...
 )
 
 const version = "1.1.0"
 
 // config vars
 var (
-	dbURL = env.String("DATABASE_URL", "postgres:///core?sslmode=disable")
+	dbURL      = env.String("DATABASE_URL", "postgres:///core?sslmode=disable")
+	listenAddr = env.String("LISTEN", ":1999")
+	dir        = env.String("DIR", defaultDir())
```
Comment (tessr) [inline cmd/corectl/main.go]: What do you think of DATA_DIR?

File: cmd/cored/main.go [hunk 17]
```diff
@@ -143,6 +147,55 @@ func runServer() {
...
+		Handler:      secureheader.DefaultConfig,
+		ReadTimeout:  httpReadTimeout,
+		WriteTimeout: httpWriteTimeout,
+		// Disable HTTP/2 for now until the Go implementation is more stable.
+		// https://github.com/golang/go/issues/16450
+		// https://github.com/golang/go/issues/17071
+		TLSNextProto: map[string]func(*http.Server, *tls.Conn, http.Handler){},
+	}
+
+	go func() {
```
Comment (tessr) [inline cmd/cored/main.go]: 1. It moved up because we need to start listening on the raft routes (`/raft/`) before we can set up the rest of the server (e.g. check if there's a config). We can't even read from raft storage without making this new node available to raft, because raft requires a round of consensus for reads as well as writes. 

2. It sprouted a goroutine (🌱 ) because the `ListenAndServe` calls are blocking and we needed to proceed through the rest of the set up. 

I'll add comments explaining all of this.

File: cmd/cored/main.go [hunk 18]
```diff
@@ -528,3 +544,8 @@ func (w *errlog) Write(p []byte) (int, error) {
 	}
 	return len(p), nil // report success for the MultiWriter
 }
+
+func defaultDir() string {
+	// TODO(kr): something in ~/Library on darwin?
+	return filepath.Join(os.Getenv("HOME"), ".cored")
```
Comment (tessr) [inline cmd/cored/main.go]: Now that you mention it, we set up directories through the installers for both windows and mac. We should probably just stick this in the same directory?

File: database/raft/raft_test.go [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+var idCases = []struct {
+	id   uint64
+	data []byte
+}{
+	{1, []byte{0, 0, 0, 0, 0, 0, 0, 1, 0x7e, 0x43, 0x31, 0x89}},
+	{2, []byte{0, 0, 0, 0, 0, 0, 0, 2, 0x6d, 0x13, 0xc2, 0x7d}},
+}
+
+var idErrorCases = []struct {
```
Comment (kr) [inline database/raft/raft_test.go]: This could be [][]byte.

File: database/raft/raft.go [hunk 20]
```diff
@@ -0,0 +1,920 @@
...
+		}
+		sendmsg(addr, data)
+	}
+}
+
+// best effort. if it fails, oh well -- that's why we're using raft.
+func sendmsg(addr string, data []byte) {
+	resp, err := http.Post("http://"+addr+"/raft/msg", contentType, bytes.NewReader(data))
+	if err != nil {
+		log.Error(context.Background(), err)
```
Comment (kr) [inline database/raft/raft.go]: Maybe add a todo here to change this to print "warning" instead of the word "error" to the log.

File: database/raft/raft.go [hunk 21]
```diff
@@ -0,0 +1,920 @@
...
+	}
+	sv.id = x.ID
+	var raftSnap raftpb.Snapshot
+	err = decodeSnapshot(x.Snap, &raftSnap)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	ctx := context.Background()
+	sv.raftStorage.ApplySnapshot(raftSnap)
```
Comment (kr) [inline database/raft/raft.go]: We should check this error.

File: database/raft/raft.go [hunk 22]
```diff
@@ -0,0 +1,920 @@
...
+func (sv *Service) join(addr, baseURL string) error {
+	reqURL := strings.TrimRight(baseURL, "/") + "/raft/join"
+	b, _ := json.Marshal(struct{ Addr string }{addr})
+	resp, err := http.Post(reqURL, contentType, bytes.NewReader(b))
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	var x struct {
```
Comment (kr) [inline database/raft/raft.go]: Might want to factor this out into a named type since we also use this same type in serveJoin.

File: database/raft/raft.go [hunk 23]
```diff
@@ -0,0 +1,920 @@
...
+	// The actual replicated data set.
+	stateMu   sync.Mutex
+	stateCond sync.Cond
+	state     *state.State
+	done      bool
+
+	// Current log position.
+	// access only from runUpdates goroutine
+	snapIndex uint64
+	confState raftpb.ConfState
```
Comment (kr) [inline database/raft/raft.go]: Probably need to move confState (and possibly also snapIndex) under stateMu.

File: cmd/corectl/main.go (hunk 16 above)
Comment (kr) [inline cmd/corectl/main.go]: How about `CORED_DATA`?

File: cmd/corectl/main.go (hunk 15 above)
Comment (kr) [inline cmd/corectl/main.go]: Not that it's any consolation, but "Go style" makes an exception for this case:

> Code generated by the protocol buffer compiler is exempt from this rule. Human-written code is held to a higher standard than machine-written code.

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: database/raft/raft.go (hunk 23 above)
Comment (kr) [inline database/raft/raft.go]: In general, it would be good to run it once under -race and see if it warns about anything.

File: cmd/corectl/main.go (hunk 16 above)
Comment (tessr) [inline cmd/corectl/main.go]: I'm going with `CORED_DATA_DIR` because we seem to like to stick nouns on the end of our env vars

File: database/raft/raft.go (hunk 23 above)
Comment (tessr) [inline database/raft/raft.go]: I've run this as-is with `-race` and didn't get any warnings (same with when I ran the tests, too). I'll look at this, though

File: database/raft/rand.go (hunk 14 above)
Comment (kr) [inline database/raft/rand.go]: Normal people almost never use arrays in Go. (We do quite a bit because we have 256-bit hash floating around all over the place so we got used to them.) So prob the most idiomatic way to write this is `b := make([]byte, 16)`.

Comment (tessr) [inline database/raft/rand.go]: I am going to idiomaticize this and also clarify that comment.

File: database/raft/raft.go (hunk 20 above)
Comment (tessr) [inline database/raft/raft.go]: I just went ahead and did it /shruggie

**Comment by tessr** [comment]
It's a miracle! We have a Chain Core running on raft on Windows!!

screen shot 2017-03-16 at 5 20 26 pm

This is ready for another pass.

**Comment by tessr** [comment]
Note also that the raft change now requires that config changes through `corectl` happen while `cored` isn't running (since each process locks the raft wal).

**Comment by kr** [comment]
:eyes:!

**Comment by tessr** [comment]
Rebased and shrunk! PTAL

File: core/config/config.go [hunk 24]
```diff
@@ -6,22 +6,24 @@ import (
...
-	"chain/database/sql"
-	chainjson "chain/encoding/json"
+	"chain/database/raft"
 	"chain/errors"
 	"chain/protocol"
 	"chain/protocol/bc"
 	"chain/protocol/state"
 )
 
+//go:generate protoc -I. -I$CHAIN/.. --go_out=. config.proto
```
Comment (bobg) [inline core/config/config.go]: Would it be better to put this line at the top of the file? And/or to add a comment saying that this produces the code for the `Config` and `BlockSigner` types?

File: core/config/config.proto [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+	bc.ProtoHash blockchain_id = 4;
+	string generator_url = 5;
+	string generator_access_token = 6;
+	string block_hsm_url = 7;
+	string block_hsm_access_token = 8;
+	uint64 configured_at = 9;
+	string block_pub = 10;
+	repeated BlockSigner signers = 11;
+	uint32 quorum = 12;
+	uint64 max_issuance_window = 13;
```
Comment (bobg) [inline core/config/config.proto]: Let's stick with the convention of adding a unit designator to identifiers like this (so, `max_issuance_window_ms`).

File: cmd/cored/main.go [hunk 26]
```diff
@@ -196,9 +207,9 @@ func runServer() {
 	if err != nil {
 		chainlog.Fatalkv(ctx, chainlog.KeyError, err)
 	}
-	resetInDevIfRequested(db)
+	resetInDevIfRequested(db, raftDB)
 
-	conf, err := config.Load(ctx, db)
+	conf, err := config.Load(ctx, db, raftDB)
```
Comment (jbowens) [inline cmd/cored/main.go]: Do we know how we're going to migrate existing configurations to the new raft/protobuf? Maybe if we don't have a config, we should check the `config` table in PostgreSQL? If it's there, try to save it via raft?

File: core/api.go [hunk 27]
```diff
@@ -113,8 +114,8 @@ func (a *API) needConfig() func(f interface{}) http.Handler {
 	return jsonHandler
 }
 
-func (a *API) buildHandler() {
-	// Setup the muxer.
+// buildCoreHandler adds the Core API routes to a preexisting http handler.
+func (a *API) buildCoreHandler() {
```
Comment (kr) [inline core/api.go]: Maybe we can get away without renaming this method?

File: core/api.go [hunk 28]
```diff
@@ -183,18 +184,17 @@ func (a *API) buildHandler() {
...
-	handler = maxBytes(handler)
+	handler = maxBytes(handler) // TODO(tessr): consider moving this to non-core specific mux
 	handler = webAssetsHandler(handler)
 	handler = healthHandler(handler)
 	for _, l := range a.requestLimits {
 		handler = limit.Handler(handler, alwaysError(errRateLimited), l.perSecond, l.burst, l.key)
 	}
 	handler = gzip.Handler{Handler: handler}
 	handler = coreCounter(handler)
-	handler = reqid.Handler(handler)
```
Comment (kr) [inline core/api.go]: Haha, so we currently have two reqid handlers in the chain? Sure enough.

File: core/config/config.proto (hunk 25 above)
Comment (kr) [inline core/config/config.proto]: This reminds me, it would be good to double check that the generated type is actually compatible. Protoc generates json tags, but I am not 100% sure that all the conventions match (e.g. decoding hex vs base64). The BlockSigner pubkey field in particular might be problematic.

File: log/log.go [hunk 29]
```diff
@@ -47,7 +47,8 @@ const (
 
 	KeyMessage = "message" // produced by Message
 	KeyError   = "error"   // produced by Error
-	KeyStack   = "stack"   // used by Printkv to print stack on subsequent lines
+	KeyWarning = "warning"
```
Comment (kr) [inline log/log.go]: Is KeyWarning unused?

File: cmd/cored/main.go (hunk 26 above)
Comment (tessr) [inline cmd/cored/main.go]: Oof, yeah, this is a good point. I've been trying to punt on "polishing" work but this will break systems with the config in postgres, so my inclination is to try to squeak that into this PR

File: log/log.go (hunk 29 above)
Comment (tessr) [inline log/log.go]: Yes, good catch

File: core/api.go (hunk 27 above)
Comment (tessr) [inline core/api.go]: yeah, fair

**Comment by tessr** [comment]
Hey guess what!!!!!!!

This is ready for another pass! (13 days later)

**Comment by kr** [comment]
:eyes:

File: cmd/corectl/main.go [hunk 30]
```diff
@@ -8,24 +8,30 @@ import (
...
 	"chain/generated/rev"
 	"chain/log"
+	"chain/protocol/bc"
 )
 
 // config vars
 var (
-	dbURL = env.String("DATABASE_URL", "postgres:///core?sslmode=disable")
+	dbURL      = env.String("DATABASE_URL", "postgres:///core?sslmode=disable")
+	listenAddr = env.String("LISTEN", ":1999")
```
Comment (kr) [inline cmd/corectl/main.go]: Maybe I'm not totally up to date on the thinking here. Somehow I thought the plan was to have corectl act as a client to do config and reset operations. I'm pretty sure we already expose the necessary RPCs.

Making corectl actually do raft raises questions like, it'll have to join the cluster to do its operation, but then does it remove itself at the end? Or will the user have to do that as a separate step?

File: core/config/config.go [hunk 31]
```diff
@@ -2,6 +2,10 @@
 // Chain Core.
 package config
 
+// The following go:generate command is included in order to generate code for
```
Comment (kr) [inline core/config/config.go]: mild suggestion:

s/The following go:generate command is included in order to//

File: core/config/config.proto [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+	bool is_generator = 3;
+	bc.ProtoHash blockchain_id = 4;
+	string generator_url = 5;
+	string generator_access_token = 6;
+	string block_hsm_url = 7;
+	string block_hsm_access_token = 8;
+	uint64 configured_at = 9;
+	string block_pub = 10;
+	repeated BlockSigner signers = 11;
+	uint32 quorum = 12;
```
Comment (kr) [inline core/config/config.proto]: Should this be `int` to avoid the type conversion on config.go line 197?

File: core/config/config.go [hunk 33]
```diff
@@ -38,31 +44,47 @@ var (
...
+		if config != nil {
+			val, err := proto.Marshal(config)
+			if err != nil {
+				return nil, errors.Wrap(err)
+			}
+			err = rDB.Insert(ctx, "/core/config", val)
+			if err != nil {
+				return nil, errors.Wrap(err)
+			}
+			return config, nil
```
Comment (kr) [inline core/config/config.go]: Should this migration logic also delete the old config to avoid confusion? Maybe it should even drop the table???

File: core/config/config.proto (hunk 32 above)
Comment (tessr) [inline core/config/config.proto]: Do protobufs have an `int` type? (I'm pretty sure they don't.)

File: cmd/corectl/main.go (hunk 30 above)
Comment (tessr) [inline cmd/corectl/main.go]: My original plan had been to turn `corectl` into a client as a polish step, since I'm reluctant to pile more into this PR, and running this seems to work okay. BUT given that I was also thinking that the ability to remove a node from a cluster is also a polish step, I guess I should fix this (that is, turn corectl into a client) in this PR.

File: core/config/config.proto (hunk 32 above)
Comment (kr) [inline core/config/config.proto]: Oh I guess not, my bad. 😅

**Comment by tessr** [comment]
This is ready for another pass 🏁

**Comment by kr** [comment]
:eyes:

**Comment by kr** [comment]
ok, actually :eyes: now

File: wercker.yml [hunk 34]
```diff
@@ -1,4 +1,4 @@
-box: chaindev/ci:20170306
+box: chaindev/ci:20170405
```
Comment (kr) [inline wercker.yml]: Is this a real diff?

**Comment by kr** [comment]
LGTM, but this should prob wait for #850 (so no more ProtoHash).

File: wercker.yml (hunk 34 above)
Comment (tessr) [inline wercker.yml]: yes.

Comment (tessr) [inline wercker.yml]: although, looks like I'll have to build a new docker image and update this again, since there's a merge conflict.
